pythonPackages.cx_oracle: fix dlopen, patchelf and fix library path for oracle-instantclient binaries

flokli opened the pull request:
###### Motivation for this change

Backport of #40688 to release-18.03.

This PR:
 - cleans up the oracle-instantclient expression
 - prefixes `LD_LIBRARY_PATH` for binaries in there
 - `patchelf`s some previously missed binaries
 - bumps pythonPackages.cx_oracle package
 - patches the `odpi` library, so it can find the oracle-instantclient libraries without having to set `ORACLE_HOME`

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
pythonPackages.cx_oracle: init at 6.1

(cherry picked from commit c9366518fa710258e8f4c3810761b08fabe47caf)

FRidh committed:
python: cx_Oracle: 6.1 -> 6.2.1

(cherry picked from commit c1932f5156b3adc81bbee872608fb898fd49cb7a)

flokli committed:
oracle-instantclient: move rpmextract and patchelf to nativeBuildInputs

(cherry picked from commit 276f2897cb567593ae7f071d13bdeec5fad61240)

flokli committed:
oracle-instantclient: remove unused arguments

(cherry picked from commit 0f4269bf48458bcc9ec089364e9dd234585f7667)

flokli committed:
oracle-instantclient: prefix LD_LIBRARY_PATH

(cherry picked from commit ccb4558a91f6466eb2aee4caeb8cb6dd42b8a082)

flokli committed:
oracle-instantclient: patchelf and wrap adrci and genezi, too

(cherry picked from commit 7824c60a439810bc068db5d67756161ac62527fa)

flokli committed:
pythonPackages.cx_oracle: 6.2.1 -> 6.3.1

(cherry picked from commit a88354d99f8298a45a2e5e3632fb793a83f2b940)

flokli committed:
pythonPackages.cx_oracle: fix dlopen

previously, $ORACLE_HOME had to be set for each python script using this
library.

We now patch odpi to load libclntsh.so from oracle-instantclient if
$ORACLE_HOME was not provided.

(cherry picked from commit 639f7952be31f89d22512bf661d03a4fe2b075a2)

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=fd058bc9-683f-43a5-bcf1-315d13650b15)

Attempted: python

The following builds were skipped because they don't evaluate on aarch64-linux: oracle-instantclient

<details><summary>Partial log (click to expand)</summary><p>

```
a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/pqby5j0dwxz3bzh0gpqlfrjvvb614r2b-python-2.7.14
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=f92660a5-eae2-4797-b00f-873034406259)

Attempted: python

The following builds were skipped because they don't evaluate on x86_64-linux: oracle-instantclient

<details><summary>Partial log (click to expand)</summary><p>

```
a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnfree = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnfree = true; }
to ~/.config/nixpkgs/config.nix.


/nix/store/nx3jw576gqw01iiijgsav39w2qa4cni2-python-2.7.14
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=f6ad37dd-ec15-400f-b750-1dcc59b441b8)

Attempted: python

The following builds were skipped because they don't evaluate on x86_64-darwin: oracle-instantclient

<details><summary>Partial log (click to expand)</summary><p>

```


these paths will be fetched (8.81 MiB download, 49.77 MiB unpacked):
  /nix/store/hldbiyiqd85xs8i2vjxccd23aj5by0f7-python-2.7.14
  /nix/store/m1zibc42jw94g2n6askrdkha96mgwp2k-gdbm-1.14
  /nix/store/pkbcpy0r299xw5ibwa7ym6giab5ys021-openssl-1.0.2o
copying path '/nix/store/m1zibc42jw94g2n6askrdkha96mgwp2k-gdbm-1.14' from 'https://cache.nixos.org'...
copying path '/nix/store/pkbcpy0r299xw5ibwa7ym6giab5ys021-openssl-1.0.2o' from 'https://cache.nixos.org'...
copying path '/nix/store/hldbiyiqd85xs8i2vjxccd23aj5by0f7-python-2.7.14' from 'https://cache.nixos.org'...
/nix/store/hldbiyiqd85xs8i2vjxccd23aj5by0f7-python-2.7.14
```
</p></details>

flokli committed:
oracle-instantclient: add flokli as maintainer

(cherry picked from commit 82eaa5ba375f98fe582db080392b2f8288528660)

flokli committed:
oracle-instantclient: simplify, use autoPatchelfHook

Split `buildCommand`, provide `unpackCmd` and add `installPhase`.

Use autoPatchelfHook, we can get rid of all the manual hacking around
with patchelf.

Use install to install to $out

(cherry picked from commit fe56ad70f0d6b38811fc2a8e2eeeb98047cac2bd)

flokli committed:
oracle-instantclient: disable building on hydra

we requireFile, so hydra can't build it anyways.

(cherry picked from commit 9770699ebd11bb5235a59d1f247faea3f9e601b1)

flokli commented:
I added the commits from #40900 into this PR as well, as this makes the expression much nicer :-)

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=7ff45d41-0d55-4947-9bd8-279ab10a4665)

Attempted: python

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/hldbiyiqd85xs8i2vjxccd23aj5by0f7-python-2.7.14
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=cb11feca-a559-4091-a580-c2862bfbed28)

Attempted: python

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/nx3jw576gqw01iiijgsav39w2qa4cni2-python-2.7.14
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.40695&attempt_id=6521eb03-9a14-4b35-9e5c-aa60815930c9)

Attempted: python

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/pqby5j0dwxz3bzh0gpqlfrjvvb614r2b-python-2.7.14
```
</p></details>

y0no commented:
:+1:
